src/FoxyClient.php

```diff
@@ -0,0 +1,637 @@
+    {
+        if ($this->accept_content_type == '') {
+            return static::DEFAULT_ACCEPT_CONTENT_TYPE;
+        }
+
+        return $this->accept_content_type;
+    }
+
+    // Get a link out of the internally stored links
+    // The link relationship base uri can be excluded ("fx:" for HAL, full URI for Siren)
+
+    public function setAcceptContentType(string $accept_content_type): void
+    {
+        $this->accept_content_type = $accept_content_type;
+    }
+
+    // Get stored links (excluding link rel base uri)
+
+    public function obtainingTokenDone(): void
+    {
+        $this->obtaining_updated_access_token = false;
+        $this->include_auth_header = true;
+    }
+
+    //Return any errors that exist in the response data.
+
+    public function getApiHome(): string
+    {
+        return $this->api_home;
+    }
+
+    public function getHeaders(): array
+    {
+        $headers = [
+            'FOXY-API-VERSION' => 1
+        ];
+
+        if ($this->access_token && $this->include_auth_header) {
+            $headers['Authorization'] = "Bearer " . $this->access_token;
+        }
+
+        $headers['Accept'] = $this->getAcceptContentType();
+
+        return $headers;
+    }
+
+    // Set a custom supported content type (application/hal+json, application/vnd.siren+json, etc)
+
+    /**
+     * @throws ClientExceptionInterface
+     * @throws JsonException
+     */
+    private function processRequest(
+        string $method,
+        string $uri,
+        ?array $post,
+        array $options
+    ): array {
+        // special case for PATCHing a Downloadable File
+        if (is_array($post) && array_key_exists('file', $post) && $method == 'PATCH') {
+            $method = 'POST';
+            $options['headers']['X-HTTP-Method-Override'] = 'PATCH';
+        }
+
+        $request = $this->requestFactory->createRequest($method, $uri);
+
+        if (!empty($options['query'])) {
+            $uri = $request->getUri();
+            $uri = $uri->withQuery(http_build_query($options['query']));
+            $request = $request->withUri($uri);
+        }
+
+        if (isset($options['form_params'])) {
+            $body = $request->getBody();
+            $body->write(http_build_query($options['form_params'], '', '&'));
+            $request = $request->withBody($body);
+        }
+
+        if (is_array($options['headers']) && !empty($options['headers'])) {
+            foreach ($options['headers'] as $headerName => $headerValue) {
+                $request = $request->withAddedHeader($headerName, $headerValue);
+            }
+        }
+
+        $response = $this->client->sendRequest($request);
+        $this->last_response = $response;
+
+        if ($response->getStatusCode() >= 400) {
+            throw new ResponseException('Error completing request', $request, $response);
+        }
+
+        $data = json_decode($response->getBody()->getContents(), true);
+
+        if ($data === null) {
```

Comment:
Checking null returned by json_decode() can be risky in some rare occasion if the response contains some binary chars. Even if the response is in correct json form.

@brettflorio

Reply:
thx @Omeryl, I wasn't aware of `JSON_THROW_ON_ERROR` flag.